tsl/src/fdw/deparse.c

```diff
@@ -867,6 +867,59 @@ deparseDistinctClause(StringInfo buf, deparse_expr_cxt *context)
 			return;
 	}
 
+	if (query->hasDistinctOn)
+	{
+        /*
+         * Pushing down DISTINCT ON is more complex than plain DISTINCT.
+         * The former requires a matching ORDER BY clause, or else the result is
+         * undefined. There is a diagnostic for the case where the ORDER BY
+         * clause doesn't match the DISTINCT ON clause, so in this case we would
+         * get an error on the data node. There is no diagnostic for the case
+         * where the ORDER BY is absent, so in this case we would get a wrong
+         * result.
+         * The remote ORDER BY clause is created from the pathkeys of the
+         * corresponding relation. If these pathkeys do not match the DISTINCT ON
+         * columns, we cannot push it down.
```

Comment:
Note that we typically add both ordered and unordered paths, so there should be a case where we can push down the DISTINCT ON and then do s Sort above it, either on the DNs or the AN.

I think this is about costing, and I think there is currently no costing implemented for pushing down DISTINCT and DISTINCT ON. This seems to be missing.

Pusing down DISTINCT means we retrieve less tuples over the network, which should always be a huge cost saving (unless all tuples are already distinct). I think we need to account for this in cost estimation when we create paths, and then let the planner decide when it makes sense to use the push-down paths or the non-push down paths.

Reply:
If it's _always_ beneficial (which I agree on), we don't need the costing and should just do it unconditionally.